File: src/helium/ModConfig.java
package helium;

import arc.files.Fi;
import arc.util.Log;
import arc.util.serialization.Jval;

import java.io.*;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import java.lang.reflect.Array;
import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.util.Arrays;

public class ModConfig {
  private static final int configVersion = 0;
  private static final Field[] configs = ModConfig.class.getFields();

  private String lastContext;

  private final Fi configFile;
  private final Fi configBack;
  private final Fi internalConfigFile;

  @Order(0) public boolean loadInfo;
  @Order(1) public boolean enableBlur;
  @Order(2) public int blurLevel;
  @Order(3) public float backBlurLen;

  public ModConfig(Fi configDir, Fi internalSource){
    configFile = configDir.child("mod_config.hjson");
    configBack = configDir.child("mod_config.hjson.bak");
    internalConfigFile = internalSource;
  }

  public void load(){
    if(!configFile.exists()){
      internalConfigFile.copyTo(configFile);
      Log.info("Configuration file is not exist, copying the default configuration");
      load(configFile);
    }
    else{
      if(!load(configFile)){
        Fi backup;
        configFile.copyTo(backup = configBack);
        internalConfigFile.copyTo(configFile);
        Log.info("default configuration file version updated, eld config should be override(backup file for old file was created)");
        load(configFile);
        String tmp = lastContext;
        load(backup, true);
        lastContext = tmp;

        save();
      }
    }

    if(loadInfo) printConfig();
  }

  public void printConfig(){
    StringBuilder results = new StringBuilder();

    for(Field cfg: configs){
      try{
        results.append("  ").append(cfg.getName()).append(" = ").append(cfg.get(this)).append(";").append(System.lineSeparator());
      }catch(IllegalAccessException e){
        throw new RuntimeException(e);
      }
    }

    Log.info("Mod config loaded! The config data:[" + System.lineSeparator() + results + "]");
  }

  public boolean load(Fi file){
    return load(file, false);
  }

  public boolean load(Fi file, boolean loadOld){
    int n;
    char[] part = new char[8192];
    StringBuilder sb = new StringBuilder();
    try(Reader r = file.reader()){
      while((n = r.read(part, 0, part.length)) != -1){
        sb.append(part, 0, n);
      }
    }catch(IOException e){
      throw new RuntimeException(e);
    }

    lastContext = sb.toString();
    Jval config = Jval.read(lastContext);

    boolean old = config.get("configVersion").asInt() != configVersion;

    if(!loadOld && old) return false;

    for(Field cfg: configs){
      if(!config.has(cfg.getName())) continue;

      String temp = config.get(cfg.getName()).toString();
      try{
        cfg.set(this, warp(cfg.getType(), temp));
      }
      catch(IllegalArgumentException | IllegalAccessException e){
        Log.err(e);
      }
    }

    return !old;
  }

  public void save(){
    try{
      save(configFile);
    }catch(IOException e){
      Log.err(e);
    }
  }

  @SuppressWarnings({"HardcodedFileSeparator", "unchecked"})
  public void save(Fi file) throws IOException{
    Jval tree = Jval.newObject();

    Jval.JsonMap map = tree.asObject();
    map.put("configVersion", Jval.valueOf(configVersion));

    Field[] configs = ModConfig.class.getFields();
    Arrays.sort(configs, (f1, f2) -> {
      float f = f1.getAnnotation(Order.class).value() - f2.getAnnotation(Order.class).value();
      if(f == 0) return 0;
      return f > 0? 1: -1;
    });
    try{
      for(Field cfg: configs){
        String key = cfg.getName();
        Object obj = cfg.get(this);
        if(obj == null){
          if(CharSequence.class.isAssignableFrom(cfg.getType())){
            map.put(key, Jval.valueOf(""));
          }
          else if(cfg.getType().isArray()){
            map.put(key, Jval.newArray());
          }
          else if(cfg.getType().isEnum()){
            map.put(key, Jval.valueOf(firstEnum((Class<? extends Enum<?>>) cfg.getType()).name()));
          }
        }
        else map.put(key, pack(obj));
      }
    }catch(IllegalAccessException e){
      throw new RuntimeException(e);
    }

    StringWriter writer = new StringWriter();
    tree.writeTo(writer, Jval.Jformat.formatted);

    String str = writer.getBuffer().toString();
    BufferedReader r1 = new BufferedReader(new StringReader(str));
    BufferedReader r2 = new BufferedReader(new StringReader(lastContext));

    BufferedWriter write = new BufferedWriter(file.writer(false));
    
    String line;
    while((line = r2.readLine()) != null){
      int i;
      String after = "";
      if((i = line.indexOf("//")) != -1){
        if(line.substring(0, i).trim().isEmpty()){
          write.write(line);
        }
        else after = line.substring(i);
      }
      else if(line.isEmpty()){
        write.write("");
      }

      if(!line.isEmpty() && (i == -1 || !after.equals(""))) write.write(r1.readLine());

      write.write(after);
      
      write.write(System.lineSeparator());
      write.flush();
    }
    write.close();
    r1.close();
    r2.close();
  }

  private static Jval pack(Object value){
    Class<?> type = value.getClass();
    if(type == Integer.class) return Jval.valueOf((int)value);
    else if(type == Byte.class) return Jval.valueOf((byte)value);
    else if(type == Short.class) return Jval.valueOf((short)value);
    else if(type == Boolean.class) return Jval.valueOf((boolean)value);
    else if(type == Long.class) return Jval.valueOf((long)value);
    else if(type == Character.class) return Jval.valueOf((char)value);
    else if(type == Float.class) return Jval.valueOf((float)value);
    else if(type == Double.class) return Jval.valueOf((double)value);
    else if(CharSequence.class.isAssignableFrom(type)) return Jval.valueOf((String) value);
    else if(type.isArray()) return packArray(value);
    else if(type.isEnum()) return Jval.valueOf(((Enum<?>)value).name());
    else throw new RuntimeException("invalid type: " + type);
  }

  private static Jval packArray(Object array){
    if(!array.getClass().isArray()) throw new RuntimeException("given object was not an array");

    int len = Array.getLength(array);
    Jval res = Jval.newArray();
    Jval.JsonArray arr = res.asArray();
    for(int i = 0; i < len; i++){
      arr.add(pack(Array.get(array, i)));
    }

    return res;
  }

  @SuppressWarnings({"unchecked", "rawtypes"})
  private static <T> T warp(Class<T> type, String value){
    if(type == int.class) return (T)Integer.valueOf(value);
    else if(type == byte.class) return (T)Byte.valueOf(value);
    else if(type == short.class) return (T)Short.valueOf(value);
    else if(type == boolean.class) return (T)Boolean.valueOf(value);
    else if(type == long.class) return (T)Long.valueOf(value);
    else if(type == char.class) return (T)Character.valueOf(value.charAt(0));
    else if(type == float.class) return (T)Float.valueOf(value);
    else if(type == double.class) return (T)Double.valueOf(value);
    else if(CharSequence.class.isAssignableFrom(type)) return (T) value;
    else if(type.isArray()) return toArray(type, value);
    else if(type.isEnum()) return (T) Enum.valueOf((Class) type, value);
    else throw new RuntimeException("invalid type: " + type);
  }

  @SuppressWarnings("unchecked")
  private static <T> T toArray(Class<T> type, String value){
    if(!type.isArray()) throw new RuntimeException("class " + type + " was not an array");
    Jval.JsonArray a = Jval.read(value).asArray();
    Class<?> eleType = type.getComponentType();
    Object res = Array.newInstance(eleType, a.size);
    for(int i = 0; i < a.size; i++){
      Array.set(res, i, warp(eleType, a.get(i).toString()));
    }

    return (T) res;
  }

  @SuppressWarnings("unchecked")
  private static <T> T firstEnum(Class<T> type){
    if(!type.isEnum()) throw new RuntimeException("class " + type + " was not an enum");
    try {
      return ((T[]) type.getMethod("values").invoke(null))[0];
    } catch (IllegalAccessException|InvocationTargetException|NoSuchMethodException e) {
      throw new RuntimeException(e);
    }
  }

  public void reset() {
    configFile.copyTo(configBack);
    configFile.delete();

    Log.info("[Singularity][INFO] mod config has been reset, old config file saved to file named \"mod_config.hjson.bak\"");
    load();
  }

  @Retention(RetentionPolicy.RUNTIME)
  @Target({ElementType.FIELD})
  private @interface Order{
    float value();
  }
}

File: src/main/kotlin/helium/ModConfig.kt
package helium

import arc.files.Fi
import arc.util.Log
import arc.util.serialization.Jval
import java.io.BufferedReader
import java.io.IOException
import java.io.StringReader
import java.io.StringWriter

private typealias RArray = java.lang.reflect.Array

class ModConfig(configDir: Fi, internalSource: Fi) {
  companion object {
    private const val configVersion = 0
    private val configs = ModConfig::class.java.declaredFields
      .filter { it.getAnnotation(ConfigItem::class.java) != null }
      .sortedBy { it.getAnnotation(ConfigItem::class.java).order }
  }

  private var lastContext: String? = null

  private val configFile: Fi = configDir.child("mod_config.hjson")
  private val configBack: Fi = configDir.child("mod_config.hjson.bak")
  private val internalConfigFile: Fi = internalSource

  @ConfigItem(order = 0) var loadInfo = false
  @ConfigItem(order = 1) var enableBlur = false
  @ConfigItem(order = 2) var blurScl = 0
  @ConfigItem(order = 3) var blurSpace = 0f

  @ConfigItem(order = 4) var entityInfoFlushInterval = 0f
  @ConfigItem(order = 5) var entityInfoLimit = 0
  @ConfigItem(order = 6) var entityInfoScale = 0f

  fun load() {
    if (!configFile.exists()) {
      internalConfigFile.copyTo(configFile)
      Log.info("Configuration file is not exist, copying the default configuration")
      load(configFile)
    } else {
      if (!load(configFile)) {
        val backup = configBack
        configFile.copyTo(backup)
        internalConfigFile.copyTo(configFile)
        Log.info("default configuration file version updated, old config should be override(backup file for old file was created)")
        load(configFile)
        val tmp = lastContext
        load(backup, true)
        lastContext = tmp
        save()
      }
    }

    if (loadInfo) printConfig()
  }

  fun printConfig() {
    val results = StringBuilder()

    configs.forEach { cfg ->
      try {
        results.append("  ")
            .append(cfg.name)
            .append(" = ")
            .append(cfg.get(this))
            .append(";")
            .append(System.lineSeparator())
      } catch (e: IllegalAccessException) {
        throw RuntimeException(e)
      }
    }

    Log.info("Mod config loaded! The config data:[${System.lineSeparator()}$results]")
  }

  fun load(file: Fi): Boolean = load(file, false)

  fun load(file: Fi, loadOld: Boolean): Boolean {
    val sb = StringBuilder()
    file.reader().use { reader ->
      val part = CharArray(8192)
      var n: Int
      while (reader.read(part, 0, part.size).also { n = it } != -1) {
        sb.appendRange(part, 0, n)
      }
    }

    lastContext = sb.toString()
    val config = Jval.read(lastContext!!)

    val old = config.get("configVersion").asInt() != configVersion
    if (!loadOld && old) return false

    configs.forEach { cfg ->
      if (!config.has(cfg.name)) return@forEach

      val temp = config.get(cfg.name).toString()
      try {
        cfg.set(this, warp(cfg.type, temp))
      } catch (e: IllegalArgumentException) {
        Log.err(e.toString())
      } catch (e: IllegalAccessException) {
        Log.err(e.toString())
      }
    }

    return !old
  }

  fun save() {
    try {
      save(configFile)
    } catch (e: IOException) {
      Log.err(e.toString())
    }
  }

  @Suppress("UNCHECKED_CAST")
  fun save(file: Fi) {
    val tree = Jval.newObject()
    val map = tree.asObject()
    map.put("configVersion", Jval.valueOf(configVersion))

    configs.forEach { cfg ->
      val key = cfg.name
      val obj = cfg.get(this)
      when {
        obj == null -> when {
          CharSequence::class.java.isAssignableFrom(cfg.type) -> map.put(key, Jval.valueOf(""))
          cfg.type.isArray -> map.put(key, Jval.newArray())
          cfg.type.isEnum -> map.put(key, Jval.valueOf(firstEnum(cfg.type as Class<Enum<*>>).name))
          else -> throw RuntimeException("Unhandled null type")
        }
        else -> map.put(key, pack(obj))
      }
    }

    val writer = StringWriter()
    tree.writeTo(writer, Jval.Jformat.formatted)

    val str = writer.buffer.toString()
    val r1 = BufferedReader(StringReader(str))
    val r2 = BufferedReader(StringReader(lastContext!!))

    file.writer(false).use { write ->
      var line: String?
      while (r2.readLine().also { line = it } != null) {
        var i: Int
        var after = ""
        line?.let { currentLine ->
          i = currentLine.indexOf("//")
          if (i != -1) {
            if (currentLine.substring(0, i).trim().isEmpty()) {
              write.write(currentLine)
            } else {
              after = currentLine.substring(i)
            }
          } else if (currentLine.isEmpty()) {
            write.write("")
          }

          if (currentLine.isNotEmpty() && (i == -1 || after.isNotEmpty())) {
            write.write(r1.readLine())
          }

          write.write(after)
          write.write(System.lineSeparator())
          write.flush()
        }
      }
    }
    r1.close()
    r2.close()
  }

  private fun pack(value: Any): Jval {
    return when (value) {
      is Int -> Jval.valueOf(value)
      is Byte -> Jval.valueOf(value.toInt())
      is Short -> Jval.valueOf(value.toInt())
      is Boolean -> Jval.valueOf(value)
      is Long -> Jval.valueOf(value)
      is Char -> Jval.valueOf(value.code)
      is Float -> Jval.valueOf(value)
      is Double -> Jval.valueOf(value)
      is CharSequence -> Jval.valueOf(value.toString())
      else -> when {
        value.javaClass.isArray -> packArray(value)
        value.javaClass.isEnum -> Jval.valueOf((value as Enum<*>).name)
        else -> throw RuntimeException("invalid type: ${value.javaClass}")
      }
    }
  }

  private fun packArray(array: Any): Jval {
    if (!array.javaClass.isArray) throw RuntimeException("given object was not an array")

    val len = RArray.getLength(array)
    val res = Jval.newArray()
    val arr = res.asArray()
    (0 until len).forEach { i ->
        arr.add(pack(RArray.get(array, i)))
    }
    return res
  }

  @Suppress("UNCHECKED_CAST")
  private fun <T> warp(type: Class<T>, value: String): T {
    return when (type) {
      Int::class.java -> value.toInt() as T
      Byte::class.java -> value.toByte() as T
      Short::class.java -> value.toShort() as T
      Boolean::class.java -> value.toBoolean() as T
      Long::class.java -> value.toLong() as T
      Char::class.java -> value[0] as T
      Float::class.java -> value.toFloat() as T
      Double::class.java -> value.toDouble() as T
      else -> when {
        CharSequence::class.java.isAssignableFrom(type) -> value as T
        type.isArray -> toArray(type, value)
        type.isEnum -> (type as Class<Enum<*>>).enumConstants.find { it.name == value } as T
        else -> throw RuntimeException("invalid type: $type")
      }
    }
  }

  @Suppress("UNCHECKED_CAST")
  private fun <T> toArray(type: Class<T>, value: String): T {
    if (!type.isArray) throw RuntimeException("class $type was not an array")
    val a = Jval.read(value).asArray()
    val eleType = type.componentType
    val res = RArray.newInstance(eleType, a.size)
    (0 until a.size).forEach { i ->
        RArray.set(res, i, warp(eleType, a[i].toString()))
    }
    return res as T
  }

  @Suppress("UNCHECKED_CAST")
  private fun <T: Enum<*>> firstEnum(type: Class<T>): T {
    if (!type.isEnum) throw RuntimeException("class $type was not an enum")

    return (type.getMethod("values").invoke(null) as Array<T>)[0]
  }

  fun reset() {
    configFile.copyTo(configBack)
    configFile.delete()
    Log.info("[INFO][${He.MOD_NAME}] mod config has been reset, old config file saved to file named \"mod_config.hjson.bak\"")
    load()
  }
}

@Retention(AnnotationRetention.RUNTIME)
@Target(AnnotationTarget.FIELD)
annotation class ConfigItem(val order: Int)
